nsxt/resource_nsxt_upgrade_run.go

```diff
@@ -0,0 +1,568 @@
+	PlanClient    upgrade.PlanClient
+	StatusClient  upgrade.StatusSummaryClient
+	UpgradeClient nsx.UpgradeClient
+}
+
+func newUpgradeClientSet(connector client.Connector) upgradeClientSet {
+	return upgradeClientSet{
+		GroupClient:   upgrade.NewUpgradeUnitGroupsClient(connector),
+		SettingClient: plan.NewSettingsClient(connector),
+		PlanClient:    upgrade.NewPlanClient(connector),
+		StatusClient:  upgrade.NewStatusSummaryClient(connector),
+		UpgradeClient: nsx.NewUpgradeClient(connector),
+	}
+}
+
+func resourceNsxtUpgradeRun() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtUpgradeRunCreate,
+		Read:   resourceNsxtUpgradeRunRead,
+		Update: resourceNsxtUpgradeRunUpdate,
+		Delete: resourceNsxtUpgradeRunDelete,
+
+		Schema: map[string]*schema.Schema{
+			"upgrade_prepare_id": {
+				Type:        schema.TypeString,
+				Description: "ID of corresponding nsxt_upgrade_prepare resource",
+				Required:    true,
+			},
+			"edge_group":           getUpgradeGroupSchema(false),
+			"host_group":           getUpgradeGroupSchema(false),
+			"edge_upgrade_setting": getUpgradeSettingSchema(true),
+			"host_upgrade_setting": getUpgradeSettingSchema(false),
+			"upgrade_plan":         getUpgradeGroupSchema(true),
+			"state": {
+				Type:        schema.TypeList,
+				Description: "Upgrade states",
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:        schema.TypeString,
+							Description: "Component type",
+							Computed:    true,
+						},
+						"status": {
+							Type:        schema.TypeString,
+							Description: "Upgrade status of component",
+							Computed:    true,
+						},
+						"percent_complete": {
+							Type:        schema.TypeInt,
+							Description: "Indicator of upgrade progress in percentage",
+							Computed:    true,
+						},
+						"details": {
+							Type:        schema.TypeString,
+							Description: "Details about the upgrade status",
+							Computed:    true,
+						},
+						"target_version": {
+							Type:        schema.TypeString,
+							Description: "Target component version",
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func getUpgradeGroupSchema(compute bool) *schema.Schema {
+	elemSchema := map[string]*schema.Schema{
+		"upgrade_mode": {
+			Type:         schema.TypeString,
+			Description:  "Upgrade mode",
+			Optional:     true,
+			Computed:     compute,
+			ValidateFunc: validation.StringInSlice(supportedUpgradeGroupExtendedConfigValue["upgrade_mode"], false),
+		},
+		"maintenance_mode_config_vsan_mode": {
+			Type:         schema.TypeString,
+			Description:  "Maintenance mode config vsan mode",
+			Optional:     true,
+			Computed:     compute,
+			ValidateFunc: validation.StringInSlice(supportedUpgradeGroupExtendedConfigValue["maintenance_mode_config_vsan_mode"], false),
+		},
+		"maintenance_mode_config_evacuate_powered_off_vms": {
+			Type:         schema.TypeString,
```

Comment:
Why don't we make these boolean? In order to have an "unset" option?

Reply:
Yes, it's one of supported options of `extended_configuration`, so it could be "unset".

Reply:
It seems to me that default exists here, and the platform has 2 behaviors rather than 3? If that's the case, I suggest to make those booleans. We can choose to set them in extended config only if different from default.

Reply:
Make sense. Done.